File: ui/src/stores/layerStore/batch.ts
/**
 * Layer Batch Operations
 *
 * Batch operations for multiple layers: sequencing, exponential scale.
 * Extracted from layerStore.ts during Phase 1B modularization.
 *
 * @see docs/graphs/layerActions.md
 */

import type { Layer } from "@/types/project";
import { storeLogger } from "@/utils/logger";
import { markLayerDirty } from "@/services/layerEvaluationCache";
import type {
  ExponentialScaleOptions,
  SequenceLayersOptions,
} from "./types";
import { useProjectStore } from "../projectStore";

// ============================================================================
// SEQUENCE LAYERS
// ============================================================================

/**
 * Sequence layers - arrange them one after another in timeline
 * @param layerIds - Array of layer IDs to sequence
 * @param options - Sequencing options
 * @returns Number of layers sequenced
 */
export function sequenceLayers(
  layerIds: string[],
  options: SequenceLayersOptions = {},
): number {
  const projectStore = useProjectStore();
  const { gapFrames = 0, startFrame = 0, reverse = false } = options;

  if (layerIds.length < 2) {
    storeLogger.warn("sequenceLayers: need at least 2 layers");
    return 0;
  }

  // Get layers in order
  const layers = layerIds
    .map((id) =>
      projectStore.getActiveCompLayers().find((l: Layer) => l.id === id),
    )
    .filter((l: Layer | undefined): l is Layer => l !== null && l !== undefined);

  if (layers.length < 2) {
    storeLogger.warn("sequenceLayers: could not find enough layers");
    return 0;
  }

  // Optionally reverse the order
  const orderedLayers = reverse ? [...layers].reverse() : layers;

  // Push history BEFORE changes for undo support
  projectStore.pushHistory();

  let currentFrame = startFrame;

  orderedLayers.forEach((layer: Layer) => {
    const duration = layer.endFrame - layer.startFrame;

    // Set new start/end frames
    layer.startFrame = currentFrame;
    layer.endFrame = currentFrame + duration;

    // Move to next position
    currentFrame = layer.endFrame + gapFrames;

    markLayerDirty(layer.id);
  });

  projectStore.project.meta.modified = new Date().toISOString();

  storeLogger.info(
    `sequenceLayers: sequenced ${orderedLayers.length} layers starting at frame ${startFrame}`,
  );
  return orderedLayers.length;
}

// ============================================================================
// EXPONENTIAL SCALE
// ============================================================================

/**
 * Create exponential scale animation on a layer
 * Uses exponential curve instead of linear for more natural zoom effect.
 * Formula: scale(t) = startScale * (endScale/startScale)^t
 *
 * @param layerId - Layer to apply exponential scale to
 * @param options - Scale animation options
 * @returns Number of keyframes created
 */
export function applyExponentialScale(
  layerId: string,
  options: ExponentialScaleOptions = {},
): number {
  const projectStore = useProjectStore();
  const {
    startScale = 100,
    endScale = 200,
    startFrame = 0,
    endFrame = 30,
    keyframeCount = 10,
    axis = "both",
  } = options;

  const layer = projectStore
    .getActiveCompLayers()
    .find((l: Layer) => l.id === layerId);
  if (!layer) {
    storeLogger.warn("applyExponentialScale: layer not found");
    return 0;
  }

  // Validate scale values (prevent division by zero)
  if (!Number.isFinite(startScale) || startScale === 0) {
    storeLogger.warn(
      "applyExponentialScale: startScale must be a non-zero finite number",
    );
    return 0;
  }
  if (!Number.isFinite(endScale)) {
    storeLogger.warn(
      "applyExponentialScale: endScale must be a finite number",
    );
    return 0;
  }

  // Push history BEFORE changes for undo support
  projectStore.pushHistory();

  // Clear existing scale keyframes
  layer.transform.scale.keyframes = [];
  layer.transform.scale.animated = true;

  const duration = endFrame - startFrame;
  const ratio = endScale / startScale;

  // Generate keyframes with exponential interpolation
  for (let i = 0; i <= keyframeCount; i++) {
    const t = i / keyframeCount; // 0 to 1
    const frame = Math.round(startFrame + t * duration);

    // Exponential formula: startScale * ratio^t
    const scaleValue = startScale * ratio ** t;

    const currentValue = layer.transform.scale.value;
    let newValue: { x: number; y: number; z?: number };

    if (axis === "x") {
      newValue = { x: scaleValue, y: currentValue.y, z: currentValue.z };
    } else if (axis === "y") {
      newValue = { x: currentValue.x, y: scaleValue, z: currentValue.z };
    } else {
      newValue = { x: scaleValue, y: scaleValue, z: currentValue.z };
    }

    layer.transform.scale.keyframes.push({
      id: `kf_expscale_${frame}_${Date.now()}_${i}`,
      frame,
      value: newValue,
      interpolation: "linear", // Linear between exponential samples gives smooth curve
      inHandle: { frame: -2, value: 0, enabled: false },
      outHandle: { frame: 2, value: 0, enabled: false },
      controlMode: "smooth",
    });
  }

  markLayerDirty(layerId);
  projectStore.project.meta.modified = new Date().toISOString();

  storeLogger.info(
    `applyExponentialScale: created ${keyframeCount + 1} keyframes for exponential scale ${startScale}% -> ${endScale}%`,
  );
  return keyframeCount + 1;
}

File: ComfyUI-Lattice-Compositor/src/Lattice/State/Layer/Batch.hs
-- |
-- Module      : Lattice.State.Layer.Batch
-- Description : Batch operations for layers
-- 
-- Extracted from Lattice.State.Layer
-- Pure functions for batch layer operations (sequencing, exponential scale)
--
{-# LANGUAGE OverloadedStrings #-}

module Lattice.State.Layer.Batch
  ( sequenceLayers
  , applyExponentialScale
  ) where

import Data.HashMap.Strict (HashMap)
import qualified Data.HashMap.Strict as HM
import Data.List (mapAccumL)
import Data.Maybe (mapMaybe)
import Data.Maybe (Maybe(..))
import Data.Text (Text)
import qualified Data.Text as T
import Lattice.State.Layer.Queries (getLayerById)
import Lattice.State.Layer.Types
  ( SequenceLayersOptions(..)
  , ExponentialScaleOptions(..)
  , ExponentialScaleAxis(..)
  )
import Lattice.Types.Animation
  ( AnimatableProperty(..)
  , Keyframe(..)
  , PropertyType(..)
  , InterpolationType(..)
  , BaseInterpolationType(..)
  , BezierHandle(..)
  , ControlMode(..)
  )
import Lattice.Types.Layer (Layer(..))
import Lattice.Types.Primitives
  ( Vec3(..)
  , validateFinite
  )
import Lattice.Types.Transform (LayerTransform(..))

-- ============================================================================
-- BATCH OPERATIONS
-- ============================================================================

-- | Sequence layers - arrange them one after another in timeline
-- Pure function: takes layer IDs, sequencing options, and layers list
-- Returns number of layers sequenced and new layers list, or error
sequenceLayers ::
  [Text] -> -- Layer IDs to sequence
  SequenceLayersOptions -> -- Sequencing options
  [Layer] -> -- Current layers list
  Either Text (Int, [Layer]) -- Returns (number sequenced, new layers list), or error
sequenceLayers layerIds options layers =
  if length layerIds < 2
    then Left "sequenceLayers: need at least 2 layers"
    else
      let
        -- Get layers in order (filter out missing layers)
        orderedLayers = mapMaybe (\lid -> getLayerById layers lid) layerIds
        
        -- Optionally reverse the order
        finalOrderedLayers = case sequenceLayersOptionsReverse options of
          Just True -> reverse orderedLayers
          _ -> orderedLayers
      in
        if length finalOrderedLayers < 2
          then Left "sequenceLayers: could not find enough layers"
          else
            let
              gapFrames = maybe 0.0 id (sequenceLayersOptionsGapFrames options)
              startFrame = maybe 0.0 id (sequenceLayersOptionsStartFrame options)
              
              -- Sequence layers one after another
              sequenceOneLayer currentFrame layer =
                let
                  duration = layerEndFrame layer - layerStartFrame layer
                  newStartFrame = currentFrame
                  newEndFrame = currentFrame + duration
                  updatedLayer = layer
                    { layerStartFrame = newStartFrame
                    , layerEndFrame = newEndFrame
                    }
                  nextFrame = newEndFrame + gapFrames
                in
                  (nextFrame, updatedLayer)
              
              (_, updatedLayers) = mapAccumL sequenceOneLayer startFrame finalOrderedLayers
              
              -- Create map of updated layers
              updatedLayerMap = HM.fromList (map (\l -> (layerId l, l)) updatedLayers)
              
              -- Update layers list
              newLayers =
                map (\l ->
                  case HM.lookup (layerId l) updatedLayerMap of
                    Just updated -> updated
                    Nothing -> l
                )
                layers
            in
              Right (length finalOrderedLayers, newLayers)

-- | Create exponential scale animation on a layer
-- Pure function: takes layer ID, scale options, ID generators, and layers list
-- Returns number of keyframes created and new layers list, or error
-- Uses exponential curve: scale(t) = startScale * (endScale/startScale)^t
applyExponentialScale ::
  Text -> -- Layer ID
  ExponentialScaleOptions -> -- Scale animation options
  (Text -> Text) -> -- ID generator for scale property keyframes
  [Layer] -> -- Current layers list
  Either Text (Int, [Layer]) -- Returns (number of keyframes created, new layers list), or error
applyExponentialScale targetLayerId options genKeyframeId layers =
  case getLayerById layers targetLayerId of
    Nothing -> Left ("Layer not found: " <> targetLayerId)
    Just existingLayer ->
      let
        startScale = maybe 100.0 id (exponentialScaleOptionsStartScale options)
        endScale = maybe 200.0 id (exponentialScaleOptionsEndScale options)
        startFrame = maybe 0.0 id (exponentialScaleOptionsStartFrame options)
        endFrame = maybe 30.0 id (exponentialScaleOptionsEndFrame options)
        keyframeCount = maybe 10 id (exponentialScaleOptionsKeyframeCount options)
        axis = maybe ExponentialScaleAxisBoth id (exponentialScaleOptionsAxis options)
        
        -- Validate scale values (prevent division by zero)
        ratio = if startScale == 0.0 || not (validateFinite startScale) || not (validateFinite endScale)
          then Left "applyExponentialScale: startScale must be a non-zero finite number, endScale must be finite"
          else Right (endScale / startScale)
      in
        case ratio of
          Left err -> Left err
          Right ratioVal ->
            let
              duration = endFrame - startFrame
              
              -- Generate keyframes with exponential interpolation
              generateKeyframe i =
                let
                  t = fromIntegral i / fromIntegral keyframeCount  -- 0 to 1
                  frame = startFrame + t * duration
                  -- Exponential formula: startScale * ratio^t
                  scaleValue = startScale * (ratioVal ** t)
                  
                  -- Get current scale value
                  currentScale = transformScale (layerTransform existingLayer)
                  currentScaleVal = animatablePropertyValue currentScale
                  
                  -- Apply to appropriate axis
                  newScaleVal = case axis of
                    ExponentialScaleAxisBoth -> Vec3 scaleValue scaleValue (vec3Z currentScaleVal)
                    ExponentialScaleAxisX -> Vec3 scaleValue (vec3Y currentScaleVal) (vec3Z currentScaleVal)
                    ExponentialScaleAxisY -> Vec3 (vec3X currentScaleVal) scaleValue (vec3Z currentScaleVal)
                  
                  keyframeId_ = genKeyframeId ("expScale_" <> T.pack (show i))
                  
                  keyframe = Keyframe
                    { keyframeId = keyframeId_
                    , keyframeFrame = frame
                    , keyframeValue = newScaleVal
                    , keyframeInterpolation = InterpolationBase InterpolationBezier
                    , keyframeInHandle = BezierHandle (-5.0) 0.0 True
                    , keyframeOutHandle = BezierHandle 5.0 0.0 True
                    , keyframeControlMode = ControlSmooth
                    , keyframeSpatialInTangent = Nothing
                    , keyframeSpatialOutTangent = Nothing
                    }
                in
                  keyframe
              
              keyframes = map generateKeyframe [0..keyframeCount]
              
              -- Update scale property with new keyframes
              scaleProp = transformScale (layerTransform existingLayer)
              updatedScaleProp = scaleProp
                { animatablePropertyValue = animatablePropertyValue scaleProp  -- Keep current value as default
                , animatablePropertyKeyframes = Just keyframes
                }
              
              updatedTransform = (layerTransform existingLayer)
                { transformScale = updatedScaleProp
                }
              
              updatedLayer = existingLayer
                { layerTransform = updatedTransform
                }
              
              newLayers = map (\l -> if layerId l == targetLayerId then updatedLayer else l) layers
            in
              Right (length keyframes, newLayers)
